### PR TITLE
Add Azure Cognitive Search as vector store

### PR DESCRIPTION
**Pull request by ruoccofabrizio**
# Adding Azure Cognitive Search as vector store

Add Azure Cognitive Search as vector store

### CODE REVIEW

**Comment by ruoccofabrizio** [comment]
@dev2049 @pablocastro

**Comment by ruoccofabrizio** [comment]
@farzad528

File: pyproject.toml [hunk 1]
```diff
@@ -96,6 +96,7 @@ openlm = {version = "^0.0.5", optional = true}
 azure-ai-formrecognizer = {version = "^3.2.1", optional = true}
 azure-ai-vision = {version = "^0.11.1b1", optional = true}
 azure-cognitiveservices-speech = {version = "^1.28.0", optional = true}
+azure-search-documents = {version = "11.4.0a20230509004", source = "azure-sdk-dev", optional = true}
```
Comment (dev2049) [inline pyproject.toml:99]: is this an internal alpha release? don't see it on pypi and some of the imports don't work using latest/beta versions of package

Comment (ruoccofabrizio) [reply pyproject.toml:99]: @dev2049 this is the source for the package, not currently available on pypi. 

[[tool.poetry.source]]
name = "azure-sdk-dev"
url = "https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-python/pypi/simple/"
secondary = true

Comment (dev2049) [reply pyproject.toml:99]: do you now (/are you allowed to say) when there will be an official release?

Comment (ruoccofabrizio) [reply pyproject.toml:99]: sorry, cannot comment on this. @farzad528 @pablocastro comments here?

Comment (farzad528) [reply pyproject.toml:99]: By June 29th the SDK feed will be public

Comment (ruoccofabrizio) [reply pyproject.toml:99]: Hi @dev2049 is this a blocker or can we deal with it as a temporary solution? @farzad528

File: tests/integration_tests/vectorstores/test_azuresearch.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+from langchain.embeddings.openai import OpenAIEmbeddings
+from langchain.vectorstores.azuresearch import AzureSearch
+
+load_dotenv()
+
+# Azure OpenAI settings
+openai.api_type = "azure"
+openai.api_base = os.getenv("OPENAI_API_BASE", "")
+openai.api_version = "2023-03-15-preview"
```
Comment (shohei1029) [inline tests/integration_tests/vectorstores/test_azuresearch.py:16]: How about using GA'd version ""2023-05-15" instead of the preview "2023-03-15-preview"?

Comment (ruoccofabrizio) [reply tests/integration_tests/vectorstores/test_azuresearch.py]: Yep, this was done before GA.

File: docs/modules/indexes/vectorstores/examples/azuresearch.ipynb [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Load environment variables from a .env file using load_dotenv():\n",
+    "load_dotenv()\n",
+    "\n",
+    "openai.api_type = \"azure\"\n",
+    "openai.api_base = os.getenv('OPENAI_ENDPOINT')\n",
+    "openai.api_version = \"2023-03-15-preview\"\n",
```
Comment (shohei1029) [inline docs/modules/indexes/vectorstores/examples/azuresearch.ipynb:71]: Thank you! Likewise.

File: docs/modules/indexes/vectorstores/examples/azuresearch.ipynb
Comment (CameronVetter) [inline docs/modules/indexes/vectorstores/examples/azuresearch.ipynb:1]: It would be nice if this example also contained an example of how to shape a search index properly for a vector store to work with the is...  As a starting point this is what I used:
```

def create_index(embedding_length):
    payload = json.dumps({
        "name": index_name,
        "fields": [
            {
                "name": "id",
                "type": "Edm.String",
                "key": True,
                "filterable": True
            },
            {
                "name": "title",
                "type": "Edm.String",
                "searchable": True,
                "retrievable": True
            },
            {
                "name": "content",
                "type": "Edm.String",
                "searchable": True,
                "retrievable": True
            },
            {
                "name": "tag",
                "type": "Edm.String",
                "filterable": True,
                "searchable": True,
                "retrievable": True
            },
            {
                "name": "metadata",
                "type": "Edm.String",
                "filterable": True,
                "searchable": True,
                "retrievable": True
            },
            {
                "name": "content_vector",
                "type": "Collection(Edm.Single)",
                "searchable": True,
                "retrievable": True,
                "dimensions": embedding_length,
                "vectorSearchConfiguration": "my-vector-config"
            }
        ],
        "vectorSearch": {
            "algorithmConfigurations": [
                {
                    "name": "my-vector-config",
                    "kind": "hnsw",
                    "hnswParameters": {
                        "m": 4,
                        "efConstruction": 400,
                        "metric": "cosine"
                    }
                }
            ]
        },
        "semantic": {
            "configurations": [
                {
                    "name": "my-semantic-config",
                    "prioritizedFields": {
                        "titleField": {
                            "fieldName": "title"
                        },
                        "prioritizedContentFields": [
                            {
                                "fieldName": "content"
                            }
                        ],
                        "prioritizedKeywordsFields": [
                            {
                                "fieldName": "tag"
                            }
                        ]
                    }
                }
            ]
        }
    })
    headers = {
        'Content-Type': 'application/json',
        'api-key': api_key
    }

    response = requests.request("PUT", create_index_url,
                                headers=headers, data=payload)
    print(response.content)
    return response.status_code
```

File: pyproject.toml (hunk 1 above)
Comment (izzymsft) [reply pyproject.toml:99]: @dev2049 would it be blocked until the SDK is publicly available or what if it can be handled with http clients internally would that suffice? Just checking.

File: docs/modules/indexes/vectorstores/examples/azuresearch.ipynb [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    "Set up the vector store settings using environment variables:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "vector_store_address: str = os.getenv('AZURE_SEARCH_SERVICE_ENDPOINT')\n",
```
Comment (pablocastro) [inline docs/modules/indexes/vectorstores/examples/azuresearch.ipynb:92]: Since this is a notebook people are using interactively, wouldn't it be better for these to be just variables that people update to the right value when trying it out?

File: docs/modules/indexes/vectorstores/examples/azuresearch.ipynb [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+     ]
+    }
+   ],
+   "source": [
+    "# Perform a hybrid search \n",
+    "docs = vector_store.hybrid_search(query=\"What did the president say about Ketanji Brown Jackson\", k=3)\n",
+    "print(docs[0].page_content)"
+   ]
+  }
+ ],
```
Comment (pablocastro) [inline docs/modules/indexes/vectorstores/examples/azuresearch.ipynb:217]: it would be great to have one more example where we show how to use an existing index instead of creating/indexing content on the spot.

**Comment by pablocastro** [comment]
@ruoccofabrizio one more thing I just realized: you'll want to add the vector store class to the list in the __init__.py file's __all__ array where all the others are listed.

Finally, do we need to do doc updates to make sure it shows up in the vector store list in the table of contents?

File: docs/modules/indexes/vectorstores/examples/azuresearch.ipynb (hunk 5 above)
Comment (CameronVetter) [reply docs/modules/indexes/vectorstores/examples/azuresearch.ipynb:217]: I would also add an example of using this as memory long-term chat memory similar to the example at: https://python.langchain.com/en/latest/modules/memory/types/vectorstore_retriever_memory.html 

I'm finding this to be very compelling use case for ACS + Langchain.

File: langchain/vectorstores/azuresearch.py [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+    else:
+        credential = AzureKeyCredential(key)
+    index_client: SearchIndexClient = SearchIndexClient(
+        endpoint=endpoint, credential=credential
+    )
+    try:
+        index_client.get_index(name=index_name)
+    except ResourceNotFoundError:
+        # Fields configuration
+        fields = [
```
[the remark this answers is not in the document]
Comment (CameronVetter) [reply langchain/vectorstores/azuresearch.py:73]: I was wondering this as well, it seems tag and title are left empty in almost all use cases.  I assumed this was driven by the ACS pattern defining prioritized fields and maybe they aren't useful as a langchain vector store.  I think it would be nice if these fields could get filled in during insertion though, maybe with a user specified LLM to generate them, and use them as another way to find the closest match...

**Comment by levalencia** [comment]
I am trying to search for documentation on this, my client prefers to use Azure Service where available, if Cognitive Search can be used as Vector database then this would be really nice.

Any examples to share about this new feature?

**Comment by CameronVetter** [comment]
> I am trying to search for documentation on this, my client prefers to use Azure Service where available, if Cognitive Search can be used as Vector database then this would be really nice.
> 
> Any examples to share about this new feature?

This feature is in private preview currently, and was publically announced at Build a couple of weeks ago. I think it was covered in this session: https://build.microsoft.com/en-US/sessions/4ea50ba5-1410-4c45-9256-ba27ec10a443?source=/speakers/c4bf4673-0601-4590-a602-e78a4243982b

**Comment by farzad528** [comment]
https://github.com/Azure/cognitive-search-vector-pr

File: langchain/vectorstores/azuresearch.py [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+                name=FIELDS_TAG,
+                type=SearchFieldDataType.String,
+                filterable=True,
+                searchable=True,
+                retrievable=True,
+            ),
+            SearchableField(
+                name=FIELDS_METADATA,
+                type=SearchFieldDataType.String,
+                searchable=True,
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py:109]: A common pattern is to store the filename in the metadata which could be a good candidate for a searchable field. What do you think?

File: langchain/vectorstores/azuresearch.py [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+        ]
+        return docs
+
+    @classmethod
+    def from_texts(
+        cls: Type[AzureSearch],
+        texts: List[str],
+        embedding: Embeddings,
+        metadatas: Optional[List[dict]] = None,
+        azure_cognitive_search_name: str = os.getenv("AZURE_SEARCH_ENDPOINT", ""),
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py:431]: done through the langchain.utils get_from_env in few classes. Moving to use this function instead of directly reading with os.

File: langchain/vectorstores/azuresearch.py [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+        azure_cognitive_search_key: str = os.getenv("AZURE_SEARCH_ADMIN_KEY", ""),
+        index_name: Optional[str] = None,
+        **kwargs: Any,
+    ) -> AzureSearch:
+        # Name of the search index if not given
+        if not index_name:
+            index_name = uuid.uuid4().hex
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py:438]: Following what other vector stores are doing. Moved to a default "langchain-index" if an index name is not provided explicitly. Open to change.

File: langchain/vectorstores/azuresearch.py [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+        )
+        azure_search.add_texts(texts, metadatas, **kwargs)
+        return azure_search
+
+
+class AzureSearchVectorStoreRetriever(BaseRetriever, BaseModel):
+    vectorstore: AzureSearch
+    search_type: str = "similarity"
+    k: int = 4
+    score_threshold: float = 0.4
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py:454]: removed

File: langchain/vectorstores/azuresearch.py [hunk 11]
```diff
@@ -0,0 +1,484 @@
...
+                type=SearchFieldDataType.Collection(SearchFieldDataType.Single),
+                searchable=True,
+                dimensions=AZURESEARCH_DIMENSIONS,
+                vector_search_configuration="default",
+            ),
+            SearchableField(
+                name=FIELDS_TAG,
+                type=SearchFieldDataType.String,
+                filterable=True,
+                searchable=True,
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py:103]: removed

File: langchain/vectorstores/azuresearch.py [hunk 12]
```diff
@@ -0,0 +1,484 @@
...
+                {
+                    "@search.action": "upload",
+                    FIELDS_ID: key,
+                    FIELDS_TITLE: metadata.get(FIELDS_TITLE, ""),
+                    FIELDS_TAG: metadata.get(FIELDS_TAG, ""),
+                    FIELDS_CONTENT: text,
+                    FIELDS_CONTENT_VECTOR: np.array(
+                        self.embedding_function(text), dtype=np.float32
+                    ).tolist(),
+                    FIELDS_METADATA: json.dumps(metadata),
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py:212]: Agree, this is in line with other vector stores implementations but not the best.

File: langchain/vectorstores/azuresearch.py [hunk 13]
```diff
@@ -0,0 +1,484 @@
...
+            # Upload data in batches
+            if len(data) == MAX_UPLOAD_BATCH_SIZE:
+                response = self.client.upload_documents(documents=data)
+                # Check if all documents were successfully uploaded
+                if not all([r.succeeded for r in response]):
+                    raise Exception(response)
+                # Reset data
+                data = []
+        # Upload data to index
+        response = self.client.upload_documents(documents=data)
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py:225]: great catch!

File: langchain/vectorstores/azuresearch.py [hunk 14]
```diff
@@ -0,0 +1,484 @@
...
+from langchain.docstore.document import Document
+from langchain.embeddings.base import Embeddings
+from langchain.schema import BaseRetriever
+from langchain.vectorstores.base import VectorStore
+
+logger = logging.getLogger()
+
+AZURESEARCH_DIMENSIONS = int(
+    os.environ.get("AZURESEARCH_DIMENSIONS", 1536)
+)  # Default to OpenAI's ada-002 embedding model vector size
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py]: done

File: langchain/vectorstores/azuresearch.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py:66]: As discussed offline, will provide another example with a configurable Index.

File: langchain/vectorstores/azuresearch.py [hunk 15]
```diff
@@ -0,0 +1,484 @@
...
+        embedding_function: Callable,
+        semantic_configuration_name: Optional[str] = None,
+        semantic_query_language: str = "en-us",
+        **kwargs: Any,
+    ):
+        """Initialize with necessary components."""
+        # Initialize base class
+        self.embedding_function = embedding_function
+        self.client = get_search_client(
+            azure_cognitive_search_name,
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py]: good point, modified to endpoint for broader coverage.

File: langchain/vectorstores/azuresearch.py [hunk 16]
```diff
@@ -0,0 +1,484 @@
...
+        **kwargs: Any,
+    ) -> List[str]:
+        """Add texts data to an existing index."""
+        keys = kwargs.get("keys")
+        ids = []
+        # Write data to index
+        data = []
+        for i, text in enumerate(texts):
+            # Use provided key otherwise use default key
+            key = keys[i] if keys else str(uuid.uuid4())
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py:178]: done, thanks

File: langchain/vectorstores/azuresearch.py [hunk 17]
```diff
@@ -0,0 +1,484 @@
...
+            search_text="",
+            vector=Vector(
+                value=np.array(
+                    self.embedding_function(query), dtype=np.float32
+                ).tolist(),
+                k=k,
+                fields=FIELDS_CONTENT_VECTOR,
+            ),
+            select=[f"{FIELDS_TITLE},{FIELDS_CONTENT},{FIELDS_METADATA}"],
+            filter=filters,
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py:269]: We'll provide an example of filter creation and link to doc.

File: langchain/vectorstores/azuresearch.py [hunk 18]
```diff
@@ -0,0 +1,484 @@
...
+                    page_content=result[FIELDS_CONTENT],
+                    metadata=json.loads(result[FIELDS_METADATA]),
+                ),
+                1 - float(result["@search.score"]),
+            )
+            for result in results
+        ]
+        return docs
+
+    def hybrid_search(self, query: str, k: int = 4, **kwargs: Any) -> List[Document]:
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py]: definitely!

File: langchain/vectorstores/azuresearch.py [hunk 19]
```diff
@@ -0,0 +1,484 @@
...
+                    page_content=result[FIELDS_CONTENT],
+                    metadata=json.loads(result[FIELDS_METADATA]),
+                ),
+                1 - float(result["@search.score"]),
+            )
+            for result in results
+        ]
+        return docs
+
+    def semantic_hybrid_search(
```
[the remark this answers is not in the document]
Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py:338]: done in both :)

File: docs/modules/indexes/vectorstores/examples/azuresearch.ipynb (hunk 4 above)
Comment (ruoccofabrizio) [reply docs/modules/indexes/vectorstores/examples/azuresearch.ipynb]: sure!

**Comment by hwchase17** [comment]
> @ruoccofabrizio one more thing I just realized: you'll want to add the vector store class to the list in the **init**.py file's **all** array where all the others are listed.
> 
> Finally, do we need to do doc updates to make sure it shows up in the vector store list in the table of contents?

as long as theres a notebook, should show up automatically!

File: langchain/vectorstores/azuresearch.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (CameronVetter) [reply langchain/vectorstores/azuresearch.py:269]: Can you share something around filter creation in the meantime, I'm stumped how to apply a filter to a search through langchain...

File: langchain/vectorstores/azuresearch.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pablocastro) [reply langchain/vectorstores/azuresearch.py:89]: Fair enough. It'll be a little noisy since we'd be searching on all text of a json doc, so we might end up matching unexpected stuff. On the other hand, it does help to cases where you're searching for a file name you know about or something like that. Let's start like this, we can always adjust it.

**Comment by pablocastro** [comment]
LGTM

File: langchain/vectorstores/azuresearch.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (CameronVetter) [reply langchain/vectorstores/azuresearch.py:191]: I'm already finding in one of the projects that I really want to filter on just one of the json fields in the metadata.  This doesn't make any sense within the capabilities of ACS filtering and searching, I might suggest that it likely doesn't in some of the other implementations as well and this may be an opportunity to change this langchain pattern, but that is probably better done in a different PR, otherwise this one will never get merged.

File: langchain/vectorstores/azuresearch.py [hunk 20]
```diff
@@ -229,32 +213,45 @@ def add_texts(
...
+    def vector_search(
+        self, query: str, k: int = 50, **kwargs: Any
     ) -> List[Document]:
         """
         Returns the most similar indexed documents to the query text.
 
         Args:
             query (str): The query text for which to find similar documents.
-            k (int): The number of documents to return. Default is 4.
+            k (int): The number of documents to return. Default is 50.
```
Comment (CameronVetter) [inline langchain/vectorstores/azuresearch.py:237]: Why the change to a default of 50 documents?  This seems like it will almost always be way too many results to feed into a LLM without exceeding the max number of tokens.

Comment (pablocastro) [reply langchain/vectorstores/azuresearch.py:237]: +1. @ruoccofabrizio when I suggested that we change k to 50, I meant the "k" parameter to vector queries, and only when using hybrid search. Separately, the number we pass to the "top" query parameter should still be small in most cases (at least by default) since that's included into prompts.

Comment (ruoccofabrizio) [reply langchain/vectorstores/azuresearch.py]: fully agree, good catch. It was not meant to be modified outside of semantic search. Fixed ;)
